### PR TITLE
OBGM-469 Fix generating product code

### PR DESCRIPTION
**Pull request by awalkowiak**

### CODE REVIEW

File: grails-app/conf/application.yml [hunk 1]
```diff
@@ -338,7 +338,7 @@ openboxes:
         productCode:
             generatorType: "SEQUENCE" # IdentifierGeneratorTypeCode.SEQUENCE
             delimiter: "-"
-            format: '$\{productTypeCode}$\{delimiter}$\{sequenceNumber}' # single quote for curly braces escape character
+            format: "${productTypeCode}${delimiter}${sequenceNumber}"
```
Comment (jmiranda) [inline grails-app/conf/application.yml:341]: I was hoping it was going to be this easy.

File: grails-app/controllers/org/pih/warehouse/product/ProductController.groovy [hunk 2]
```diff
@@ -96,7 +95,7 @@ class ProductController {
         [products: product]
     }
 
-
+    @Transactional
```
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/product/ProductController.groovy:98]: Is this batchSave method being used somewhere?

File: grails-app/domain/org/pih/warehouse/product/Product.groovy [hunk 3]
```diff
@@ -48,21 +48,29 @@ import org.pih.warehouse.LocalizationUtil
  */
 class Product implements Comparable, Serializable {
 
-    def beforeInsert = {
+    def beforeInsert() {
         createdBy = AuthService.currentUser
```
Comment (jmiranda) [inline grails-app/domain/org/pih/warehouse/product/Product.groovy:52]: In the Shipment domain I changed the convention to include updatedBy in the insert as well. I had assumed we were already doing this everywhere else but it looks like I was wrong.

Topics for tech-huddle maybe: 
- Do you think that should be applied everywhere? 
- If so do you think we can move this to a generic solution 
- Auditable abstract class or trait (assuming that's what traits are for) that adds these methods to domains that have these audit fields?

The first of these could be done during MVP1, but the other points should be MVP2/post-migration.

Now that we might have more consistency with the Grails 3 dirty properties API we could also try to make these persistence events more fine-grained. There are only a few properties on Product that should cause a refresh to InventorySnapshot. Same with other domains for ProductAvailability.

File: grails-app/services/org/pih/warehouse/core/IdentifierService.groovy [hunk 4]
```diff
@@ -326,9 +326,9 @@ class IdentifierService {
      *          - empty productIdentifierFormat (will be generated basing on the openboxes.identifier.product.format)
      * */
     String generateProductIdentifier(ProductType productType) {
-        if (productType?.id == ConfigurationHolder.config.openboxes.identifier.defaultProductType.id) {
+        if (productType?.id == Holders.config.openboxes.identifier.defaultProductType.id) {
             // If product type is systems default then generate basing on the generator type
-            def generatorTypeConfig = ConfigurationHolder.config.openboxes.identifier.productCode.generatorType
+            def generatorTypeConfig = Holders.config.openboxes.identifier.productCode.generatorType
```
Comment (jmiranda) [inline grails-app/services/org/pih/warehouse/core/IdentifierService.groovy:331]: Do we have defaults for these in application.yml? 

We had an issue on the Dominica server a few weeks ago (I think I finally fixed it last week). It feels like we should discuss the expected vs actual behavior for the identifier config as part of a tech huddle.

File: grails-app/controllers/org/pih/warehouse/product/ProductController.groovy (hunk 2 above)
Comment (jmiranda) [reply grails-app/controllers/org/pih/warehouse/product/ProductController.groovy:98]: I just realized that you're just applying @Transactional to all methods so you can ignore this comment.

File: grails-app/controllers/org/pih/warehouse/product/ProductController.groovy [hunk 5]
```diff
@@ -185,13 +184,13 @@ class ProductController {
...
 
-        if (!productInstance.hasErrors() && productInstance.save(flush: true)) {
+        if (!productInstance.hasErrors() && productService.saveProduct(productInstance)) {
             log.info("saved product " + productInstance.errors)
             flash.message = "${warehouse.message(code: 'default.created.message', args: [warehouse.message(code: 'product.label', default: 'Product').decodeHTML(), format.product(product: productInstance).decodeHTML()])}"
             sendProductCreatedNotification(productInstance)
         }
 
-        render(view: "edit", model: [productInstance: productInstance, rootCategory: productService.getRootCategory(), locationInstance: location])
+        redirect(action: "edit", params: [id: productInstance.id])
```
Comment (jmiranda) [inline grails-app/controllers/org/pih/warehouse/product/ProductController.groovy:193]: What about when errors occur?

**Review by jmiranda — COMMENTED** [review]
Nothing specific to address other than maybe thinking about the render -> redirect change. Not sure if that's the behavior we want for the unhappy case.

File: grails-app/controllers/org/pih/warehouse/product/ProductController.groovy (hunk 5 above)
Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/product/ProductController.groovy:193]: You're right... I'll restore this

File: grails-app/services/org/pih/warehouse/core/IdentifierService.groovy (hunk 4 above)
Comment (awalkowiak) [reply grails-app/services/org/pih/warehouse/core/IdentifierService.groovy:331]: Yeah, it's SEQUENCE by default, needs to be changed according to needs. But I believe now everything works fine in that matter (in grails 1 instances)

File: grails-app/controllers/org/pih/warehouse/product/ProductController.groovy (hunk 5 above)
Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/product/ProductController.groovy]: fixed

**Review by kchelstowski — APPROVED** [review]

**Review by jmiranda — APPROVED** [review]

**Pull request merged**